llvm/include/llvm/Analysis/IndirectCallVisitor.h

```diff
@@ -19,20 +19,38 @@ namespace llvm {
 // Visitor class that finds all indirect call.
 struct PGOIndirectCallVisitor : public InstVisitor<PGOIndirectCallVisitor> {
   std::vector<CallBase *> IndirectCalls;
+  std::vector<Instruction *> VTableAddrs;
   PGOIndirectCallVisitor() = default;
 
   void visitCallBase(CallBase &Call) {
     if (Call.isIndirectCall())
-      IndirectCalls.push_back(&Call);
+      if (Call.isIndirectCall()) {
+        IndirectCalls.push_back(&Call);
+
+        LoadInst *LI = dyn_cast<LoadInst>(Call.getCalledOperand());
```

Comment:
Looking for vtables this way seems like it'll also capture cases where function pointers are stored/passed around which would have the same pattern: https://godbolt.org/z/sMve8fhPa.

With `-fwhole-program-vtables` each virtual callsite is tagged from the front end (https://godbolt.org/z/KvY4dW1qa) which would make correctly identifying vtables easier.

Reply:
thanks for the example. I modified the visitor to capture vtable only and updated two test cases in the 2nd commit accordingly.

Reply:
Taking a dependency on `-fwhole-program-vtables` for this does mean instrumentation fails when the flag is not enabled. Might make sense to detect this scenario and warn/message to the user that this is happening.

Reply:
It makes sense to improve usability. Updated description of option `enable-vtable-value-profiling` in this commit for this purpose.

I wonder if clang is a better place to validate user options, considering
1) fwhole-program-vtables are clang options; and IRPGO looks at type metadata / intrinsics generated by language frontend.
2) when vtable instrumentation (along with the optimization that uses it) is rolled out to be default, a simple c++ binary which doesn't use class inheritance could pick up the compiler flag change; and a warning is not expected for this simple binary.

Reply:
True, detection would need to be done in Clang. Maybe a better way to think about/document this is that this particular feature of IRPGO is enabled when `-fwhole-program-vtables` is enabled and otherwise disabled. The backend flag can still exist but more as an opt-out option.